spice_agent/inference/actions.py

```diff
@@ -21,12 +23,156 @@
 
 import torch  # noqa
 import transformers  # noqa
-from transformers import pipeline  # noqa
 from transformers.pipelines.base import PipelineException  # noqa
 
 LOGGER = logging.getLogger(__name__)
 
 
+class CompelEmbeddingsProvider:
+    """
+    Provides embeddings for a pipeline
+    """
+
+    def __init__(
+        self,
+        pipeline: DiffusionPipeline,
+        device,
+        prompt: str,
+        negative_prompt: str = "",
+    ) -> None:
+        self.pipeline = pipeline
+        self.device = device
+        self.prompt = prompt
+        self.negative_prompt = negative_prompt
+        self.compel = self._init_compel()
+
+    def _init_compel(self) -> Compel:
+        """
+        Initializes compel text embedding system based on type of pipeline.
+        """
+
+        # Compel settings consistent across compel objects
+        DEFAULT_COMPEL_ARGUMENTS = {
+            "truncate_long_prompts": False,
+            "device": self.device,
+        }
+
+        pipe = self.pipeline
+
+        # Currently, these are the only supported pipelines
+        if not isinstance(
+            pipe,
+            (
+                StableDiffusionPipeline,
+                StableDiffusionXLPipeline,
+                StableDiffusionXLImg2ImgPipeline,
+            ),
+        ):
+            message = f"Prompt embeddings are not supported for pipeline {type(pipe)}"
+            LOGGER.warn(message)
+            return None
+
+        # Check if embeddings are necessary
+        if pipe.tokenizer:
+            model_max_length = pipe.tokenizer.model_max_length
+            if (
+                len(self.prompt) <= model_max_length
+                and len(self.negative_prompt) <= model_max_length
+            ):
+                return None
+
+        # Configure compel objects for StableDiffusion pipeline
+        if isinstance(pipe, StableDiffusionPipeline):
+            return Compel(
+                tokenizer=pipe.tokenizer,
+                text_encoder=pipe.text_encoder,
+                returned_embeddings_type=ReturnedEmbeddingsType.LAST_HIDDEN_STATES_NORMALIZED,
+                requires_pooled=False,
+                **DEFAULT_COMPEL_ARGUMENTS,
+            )
+
+        # Configure compel objects for StableDiffusionXL pipelines
+        if isinstance(pipe, StableDiffusionXLPipeline):
+            return Compel(
+                tokenizer=[pipe.tokenizer, pipe.tokenizer_2],
+                text_encoder=[pipe.text_encoder, pipe.text_encoder_2],
+                returned_embeddings_type=ReturnedEmbeddingsType.PENULTIMATE_HIDDEN_STATES_NON_NORMALIZED,
+                requires_pooled=[False, True],
```

Comment:
different value from line 90?

Reply:
yep that's fine - different embeddings type are required here